scheduler-pool/src/lib.rs

```diff
@@ -0,0 +1,442 @@
+    },
+    solana_program_runtime::timings::ExecuteTimings,
+    solana_runtime::{
+        installed_scheduler_pool::{
+            InstalledScheduler, InstalledSchedulerPool, ResultWithTimings, SchedulerBox,
+            SchedulerId, SchedulerPoolArc, SchedulingContext, WaitReason,
+        },
+        prioritization_fee_cache::PrioritizationFeeCache,
+        vote_sender_types::ReplayVoteSender,
+    },
+    solana_scheduler::{SchedulingMode, WithSchedulingMode},
+    solana_sdk::transaction::SanitizedTransaction,
+    std::sync::{Arc, Mutex, Weak},
+};
+
+// SchedulerPool must be accessed via dyn by solana-runtime code, because of its internal fields'
+// types aren't available there...
+#[derive(Debug)]
+pub struct SchedulerPool {
+    schedulers: Mutex<Vec<SchedulerBox>>,
+    log_messages_bytes_limit: Option<usize>,
+    transaction_status_sender: Option<TransactionStatusSender>,
+    replay_vote_sender: Option<ReplayVoteSender>,
+    prioritization_fee_cache: Arc<PrioritizationFeeCache>,
+    weak_self: Weak<SchedulerPool>,
+}
+
+impl SchedulerPool {
+    pub fn new_dyn(
+        log_messages_bytes_limit: Option<usize>,
+        transaction_status_sender: Option<TransactionStatusSender>,
+        replay_vote_sender: Option<ReplayVoteSender>,
+        prioritization_fee_cache: Arc<PrioritizationFeeCache>,
+    ) -> SchedulerPoolArc {
+        Arc::new_cyclic(|weak_self| Self {
+            schedulers: Mutex::<Vec<SchedulerBox>>::default(),
+            log_messages_bytes_limit,
+            transaction_status_sender,
+            replay_vote_sender,
+            prioritization_fee_cache,
+            weak_self: weak_self.clone(),
+        })
+    }
+
+    pub fn self_arc(&self) -> Arc<Self> {
+        self.weak_self
+            .upgrade()
+            .expect("self-referencing Arc-ed pool")
+    }
+}
+
+impl InstalledSchedulerPool for SchedulerPool {
+    fn take_from_pool(&self, context: SchedulingContext) -> SchedulerBox {
+        assert!(!context.bank().with_scheduler());
+
+        let mut schedulers = self.schedulers.lock().expect("not poisoned");
+        // pop is intentional for filo, expecting relatively warmed-up scheduler due to having been
+        // returned recently
+        let maybe_scheduler = schedulers.pop();
+        if let Some(mut scheduler) = maybe_scheduler {
+            scheduler.replace_scheduler_context(context);
+            scheduler
+        } else {
+            Box::new(Scheduler::spawn(self.self_arc(), context))
+        }
+    }
+
+    fn return_to_pool(&self, scheduler: SchedulerBox) {
+        assert!(scheduler.scheduling_context().is_none());
+
+        self.schedulers
+            .lock()
+            .expect("not poisoned")
+            .push(scheduler);
+    }
+}
+
+// Currently, simplest possible implementation (i.e. single-threaded)
+// this will be replaced with more proper implementation...
+// not usable at all, especially for mainnet-beta
+#[derive(Debug)]
+struct Scheduler {
```

Comment:
good call. well, maybe this struct should be named like `PooledScheduler`. as you pointed out, this is unintended name conflict... and this isn't temporary. the struct is something which will be glueing/managing scheduling logic and threads together (in the future). `solana-scheduler` is meant to hold pure logic not depending on mundane world of solana-ledger (and even solana-runtime!)... so, i wonder i should also name the crate to `solana-unified-scheduler-logic` or something else to avoid ambiguity by calling things as _scheduler_ altogether... xD

Reply:
If you've got the separation in a branch, could you reference that to me? or just explain a bit more what you mean in terms of separting "pure logic" from threads. I suppose I'm lacking imagination, and am not yet sold on the idea that they need/should be separated.

Reply:
> If you've got the separation in a branch, could you reference that to me?

sure!: https://github.com/ryoqun/solana/tree/unified-scheduler-for-block-verification-wip-pre-strip

that branch is full of dirty code, though. however, it's runnable against mainnet-beta. (fully-performant code)

>  just explain a bit more what you mean in terms of separting "pure logic" from threads.

sure, again! some more reasoning:

I think crate-level separation provides good discipline over time for designing things, assuming creating crates is cheap in terms of human time. as you know, `solana-core`/`solana-runtime`/`solana-ledger` is too tangled up right now. so, I'm afraid to add another lump of code to there. and to the brand-new `solana-unified-scheduler-pool` as well.

discipline here means  well-cut and well-defined interface between and explicit and discrete set of dependencies for each respective crates.

in a say, this is kind of `crate-safety` as in `type-safety`.

also, that strict discipline make it easy to bench the scheduling logic by its own. i'll put bunch of benches in `solana-unified-scheduler-logic`.

lastly, i think scheduler logic shold be agnostic to the actual execution details up to the communicated information across the interface, but not more. in this way, logic eval and reasoning and reviewing should be easy. that will eventually leads to performant impl. (Imo).

as for the interface, we just need bunch of `crossbeam_channel::Channels`s to define such an api between crates (without any extra runtime overhead if those code would be in a single crate)

callee:

https://github.com/ryoqun/solana/blob/96013125e3c6f34d093447b5fe260fa08b34435a/scheduler/src/lib.rs#L2139-L2175

caller:

https://github.com/ryoqun/solana/blob/96013125e3c6f34d093447b5fe260fa08b34435a/scheduler-pool/src/lib.rs#L738-L755

another datapoint is that `solana-unified-scheduler`'s `Cargo.toml` is so small like this:

https://github.com/ryoqun/solana/blob/unified-scheduler-for-block-verification-wip-pre-strip/scheduler/Cargo.toml


if these explanation does make sense, i'll transcribe these as source comments in this pr or coming pr.

> If we do go down the route that they should indeed be separate, then I do like the name `solana-unified-scheduler-logic`.

gocha

Reply:
(this separate-or-not discussion is currently blocking all the renames and crate template creation).

Reply:
I think after examining the dependency graph the crate separation makes sense. Appreciate the detailed comments for justification.